File: frontend/lib/utils/date_utils.dart
import 'package:intl/intl.dart';

/// Utilitario para manipulacao de datas no fuso horario de Brasilia (America/Sao_Paulo)
/// UTC-3 (horario padrao)
class DateUtils {
  static final Duration _brasiliaOffset = const Duration(hours: -3);

  /// Retorna a data/hora atual no fuso horario de Brasilia
  static DateTime agora() {
    final utc = DateTime.now().toUtc();
    return utc.add(_brasiliaOffset);
  }

  /// Converte uma data UTC para o fuso de Brasilia
  static DateTime? paraHorarioBrasilia(DateTime? date) {
    if (date == null) return null;
    final utc = date.toUtc();
    return utc.add(_brasiliaOffset);
  }

  /// Converte uma string ISO para DateTime no fuso de Brasilia
  static DateTime? parseParaBrasilia(String? dateString) {
    if (dateString == null || dateString.isEmpty) return null;
    try {
      final parsed = DateTime.parse(dateString);
      return paraHorarioBrasilia(parsed);
    } catch (e) {
      return null;
    }
  }

  /// Converte uma data de Brasilia para UTC
  static DateTime? paraUTC(DateTime? date) {
    if (date == null) return null;
    return date.subtract(_brasiliaOffset);
  }

  /// Formata uma data para string ISO com offset de Brasilia (-03:00)
  static String? formatarParaISO(DateTime? date) {
    if (date == null) return null;
    final brasilia = paraHorarioBrasilia(date);
    if (brasilia == null) return null;

    final formatted = DateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS").format(brasilia);
    return '$formatted-03:00';
  }

  /// Formata uma data para exibicao no formato brasileiro
  /// Exemplo: "30/11/2025 14:30" ou "30/11/2025"
  static String formatarParaExibicao(DateTime? date, {bool incluirHora = true}) {
    if (date == null) return '';
    final brasilia = paraHorarioBrasilia(date);
    if (brasilia == null) return '';

    if (incluirHora) {
      return DateFormat('dd/MM/yyyy HH:mm').format(brasilia);
    }
    return DateFormat('dd/MM/yyyy').format(brasilia);
  }

  /// Formata uma data para exibicao relativa (ex: "ha 5 minutos", "ontem")
  static String formatarRelativo(DateTime? date) {
    if (date == null) return '';
    final brasilia = paraHorarioBrasilia(date);
    if (brasilia == null) return '';

    final now = agora();
    final diff = now.difference(brasilia);

    if (diff.inSeconds < 60) {
      return 'agora';
    } else if (diff.inMinutes < 60) {
      final min = diff.inMinutes;
      return 'ha $min ${min == 1 ? 'minuto' : 'minutos'}';
    } else if (diff.inHours < 24) {
      final hrs = diff.inHours;
      return 'ha $hrs ${hrs == 1 ? 'hora' : 'horas'}';
    } else if (diff.inDays == 1) {
      return 'ontem';
    } else if (diff.inDays < 7) {
      final days = diff.inDays;
      return 'ha $days dias';
    } else if (diff.inDays < 30) {
      final weeks = (diff.inDays / 7).floor();
      return 'ha $weeks ${weeks == 1 ? 'semana' : 'semanas'}';
    } else if (diff.inDays < 365) {
      final months = (diff.inDays / 30).floor();
      return 'ha $months ${months == 1 ? 'mes' : 'meses'}';
    } else {
      final years = (diff.inDays / 365).floor();
      return 'ha $years ${years == 1 ? 'ano' : 'anos'}';
    }
  }

  /// Retorna o inicio do dia atual em Brasilia (00:00:00)
  static DateTime inicioDoDia() {
    final hoje = agora();
    return DateTime(hoje.year, hoje.month, hoje.day);
  }

  /// Retorna o fim do dia atual em Brasilia (23:59:59.999)
  static DateTime fimDoDia() {
    final hoje = agora();
    return DateTime(hoje.year, hoje.month, hoje.day, 23, 59, 59, 999);
  }

  /// Verifica se uma data ja expirou (comparando com agora em Brasilia)
  static bool jaExpirou(DateTime? dataExpiracao) {
    if (dataExpiracao == null) return true;
    return agora().isAfter(paraHorarioBrasilia(dataExpiracao)!);
  }

  /// Formata duracao em minutos para exibicao legivel
  /// Exemplo: 90 -> "1h 30min"
  static String formatarDuracao(int? minutos) {
    if (minutos == null || minutos <= 0) return '';
    
    if (minutos < 60) {
      return '${minutos}min';
    }
    
    final horas = minutos ~/ 60;
    final mins = minutos % 60;
    
    if (mins == 0) {
      return '${horas}h';
    }
    return '${horas}h ${mins}min';
  }
}

File: backend/src/utils/dateUtils.js
/**
 * Utilitario para manipulacao de datas no fuso horario de Brasilia (America/Sao_Paulo)
 * UTC-3 (horario padrao) ou UTC-2 (horario de verao - atualmente nao usado no Brasil)
 */

const BRASILIA_OFFSET = -3; // UTC-3

/**
 * Retorna a data/hora atual no fuso horario de Brasilia
 * @returns {Date} Data atual em Brasilia
 */
function agora() {
  const now = new Date();
  const utc = now.getTime() + (now.getTimezoneOffset() * 60000);
  return new Date(utc + (3600000 * BRASILIA_OFFSET));
}

/**
 * Retorna a data/hora atual como string ISO no fuso de Brasilia
 * @returns {string} String ISO com offset de Brasilia
 */
function agoraISO() {
  const brasilia = agora();
  return formatarParaISO(brasilia);
}

/**
 * Converte uma data UTC para o fuso de Brasilia
 * @param {Date|string} date - Data em UTC
 * @returns {Date} Data convertida para Brasilia
 */
function paraHorarioBrasilia(date) {
  if (!date) return null;
  const d = typeof date === 'string' ? new Date(date) : date;
  if (isNaN(d.getTime())) return null;
  
  const utc = d.getTime() + (d.getTimezoneOffset() * 60000);
  return new Date(utc + (3600000 * BRASILIA_OFFSET));
}

/**
 * Converte uma data de Brasilia para UTC
 * @param {Date|string} date - Data em horario de Brasilia
 * @returns {Date} Data convertida para UTC
 */
function paraUTC(date) {
  if (!date) return null;
  const d = typeof date === 'string' ? new Date(date) : date;
  if (isNaN(d.getTime())) return null;
  
  return new Date(d.getTime() - (3600000 * BRASILIA_OFFSET));
}

/**
 * Formata uma data para string ISO com offset de Brasilia (-03:00)
 * @param {Date} date - Data a formatar
 * @returns {string} String ISO com offset
 */
function formatarParaISO(date) {
  if (!date || isNaN(date.getTime())) return null;
  
  const ano = date.getFullYear();
  const mes = String(date.getMonth() + 1).padStart(2, '0');
  const dia = String(date.getDate()).padStart(2, '0');
  const hora = String(date.getHours()).padStart(2, '0');
  const minuto = String(date.getMinutes()).padStart(2, '0');
  const segundo = String(date.getSeconds()).padStart(2, '0');
  const ms = String(date.getMilliseconds()).padStart(3, '0');
  
  return `${ano}-${mes}-${dia}T${hora}:${minuto}:${segundo}.${ms}-03:00`;
}

/**
 * Formata uma data para exibicao no formato brasileiro
 * @param {Date|string} date - Data a formatar
 * @param {boolean} incluirHora - Se deve incluir a hora
 * @returns {string} Data formatada (ex: "30/11/2025 14:30")
 */
function formatarParaExibicao(date, incluirHora = true) {
  const d = paraHorarioBrasilia(date);
  if (!d) return '';
  
  const dia = String(d.getDate()).padStart(2, '0');
  const mes = String(d.getMonth() + 1).padStart(2, '0');
  const ano = d.getFullYear();
  
  if (!incluirHora) {
    return `${dia}/${mes}/${ano}`;
  }
  
  const hora = String(d.getHours()).padStart(2, '0');
  const minuto = String(d.getMinutes()).padStart(2, '0');
  
  return `${dia}/${mes}/${ano} ${hora}:${minuto}`;
}

/**
 * Retorna o inicio do dia atual em Brasilia (00:00:00)
 * @returns {Date}
 */
function inicioDoDia() {
  const hoje = agora();
  hoje.setHours(0, 0, 0, 0);
  return hoje;
}

/**
 * Retorna o fim do dia atual em Brasilia (23:59:59.999)
 * @returns {Date}
 */
function fimDoDia() {
  const hoje = agora();
  hoje.setHours(23, 59, 59, 999);
  return hoje;
}

/**
 * Calcula expiracao adicionando dias a partir de agora (em Brasilia)
 * @param {number} dias - Numero de dias a adicionar
 * @returns {Date}
 */
function expiracaoEmDias(dias) {
  const data = agora();
  data.setDate(data.getDate() + dias);
  return data;
}

/**
 * Calcula expiracao adicionando horas a partir de agora (em Brasilia)
 * @param {number} horas - Numero de horas a adicionar
 * @returns {Date}
 */
function expiracaoEmHoras(horas) {
  const data = agora();
  data.setHours(data.getHours() + horas);
  return data;
}

/**
 * Verifica se uma data ja expirou (comparando com agora em Brasilia)
 * @param {Date|string} dataExpiracao - Data de expiracao
 * @returns {boolean}
 */
function jaExpirou(dataExpiracao) {
  if (!dataExpiracao) return true;
  const expiracao = typeof dataExpiracao === 'string' ? new Date(dataExpiracao) : dataExpiracao;
  return agora() > expiracao;
}

module.exports = {
  BRASILIA_OFFSET,
  agora,
  agoraISO,
  paraHorarioBrasilia,
  paraUTC,
  formatarParaISO,
  formatarParaExibicao,
  inicioDoDia,
  fimDoDia,
  expiracaoEmDias,
  expiracaoEmHoras,
  jaExpirou
};
